lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -255,34 +170,122 @@ Ufe::Transform3d::Ptr createTransform3d(
+
+    struct InitialUndoCalledState : public State
+    {
+        const char* name() const override { return "initial undo called"; }
+        void        handleSet(UsdTRSUndoableCmdBase* cmd, const VtValue&) override
+        {
+            // Maya triggers a redo on command creation, ignore it.
+            cmd->_state = &UsdTRSUndoableCmdBase::_initialState;
+        }
+    };
+
+    struct ExecuteState : public State
+    {
+        const char* name() const override { return "execute"; }
+        void        handleUndo(UsdTRSUndoableCmdBase* cmd) override
+        {
+            cmd->recreateOp();
+            cmd->setValue(cmd->_prevOpValue);
+            cmd->_state = &UsdTRSUndoableCmdBase::_undoneState;
+        }
+        void handleSet(UsdTRSUndoableCmdBase* cmd, const VtValue& v) override
+        {
+            cmd->_newOpValue = v;
+            cmd->setValue(v);
+        }
+    };
+
+    struct UndoneState : public State
+    {
+        const char* name() const override { return "undone"; }
+        void        handleSet(UsdTRSUndoableCmdBase* cmd, const VtValue&) override
+        {
+            // Can ignore the value, we already have it --- or assert they're
+            // equal, perhaps.
+            cmd->recreateOp();
+            cmd->setValue(cmd->_newOpValue);
+            cmd->_state = &UsdTRSUndoableCmdBase::_redoneState;
+        }
+    };
+
+    struct RedoneState : public State
+    {
+        const char* name() const override { return "redone"; }
+        void        handleUndo(UsdTRSUndoableCmdBase* cmd) override
+        {
+            cmd->recreateOp();
+            cmd->setValue(cmd->_prevOpValue);
+            cmd->_state = &UsdTRSUndoableCmdBase::_undoneState;
+        }
+    };
+
     UsdTRSUndoableCmdBase(
-        const VtValue&        newOpValue,
-        const Ufe::Path&      path,
-        const UsdGeomXformOp& op,
-        const UsdTimeCode&    writeTime_)
+        const VtValue&     newOpValue,
+        const Ufe::Path&   path,
+        OpFunc             opFunc,
+        const UsdTimeCode& writeTime_)
         : Ufe::SetVector3dUndoableCommand(path)
-        , fOp(op)
         ,
         // Always read from proxy shape time.
-        fReadTime(getTime(path))
-        , fWriteTime(writeTime_)
-        , fPrevOpValue(getValue(op.GetAttr(), readTime()))
-        , fAttrName(op.GetOpName())
-        , fNewOpValue(newOpValue)
+        _readTime(getTime(path))
+        , _writeTime(writeTime_)
+        , _prevOpValue()
+        , _newOpValue(newOpValue)
+        , _attrName()
+        , _op()
+        , _opFunc(std::move(opFunc))
     {
     }
 
-    // Have separate execute override which is value-only, as default execute()
-    // calls redo, and undo / redo will eventually not be value only and will
-    // support propertySpec / primSpec management.
-    void execute() override { setValue(fNewOpValue); }
+    // Ufe::UndoableCommand overrides.
+    void execute() override { handleSet(_newOpValue); }
```

Comment:
why do we need this handleSet method? for undo you call the state directly.

Reply:
Turns out command derived classes call this, and I don't want to expose them to the base class state, which is an implementation detail.